File: ios/GradeThread/Inventory/BulkAction.swift
import Foundation

/// One bulk action surfaced in the bottom action bar. Stage-appropriate
/// — `actions(for:)` returns the right set for the active inventory tab.
///
/// The eBay-backed actions (drop price, end listing) are wired through
/// ``BulkActionExecutor`` to the edge publish endpoints (US-185); each
/// item that has no active eBay listing fails individually with a reason
/// rather than failing the whole batch.
///
/// US-1130: the former `aiEnrich` case was removed — its executor was a
/// permanent stub that never appeared in any visible action set, so it was a
/// dead, half-advertised AI capability. Per-item AI enrichment still lives on
/// the item canvas (``AIFillReviewSheet``); there is no bulk path on iOS.
public enum BulkAction: Identifiable, Hashable {
    case createDraft
    /// US-680: validate + push each selected item to eBay in one batch.
    case publish
    case markShipped
    case endListing
    /// Percent reduction (e.g. 10 for -10%). Surfaced as a single button
    /// because that's how the web copy reads.
    case dropPrice(percent: Int)
    /// Submit the selection for certified grading. Unlike the others this
    /// doesn't run through ``BulkActionExecutor`` — the InventoryListView
    /// intercepts it to present a dedicated batch-grading sheet (tier +
    /// readiness + credits).
    case grade
    case exportCSV

    public var id: String {
        switch self {
        case .createDraft:        return "create_draft"
        case .publish:            return "publish"
        case .markShipped:        return "mark_shipped"
        case .endListing:         return "end_listing"
        case .dropPrice(let pct): return "drop_price_\(pct)"
        case .grade:              return "grade"
        case .exportCSV:          return "export_csv"
        }
    }

    public var label: String {
        switch self {
        case .createDraft:        return "Create draft"
        case .publish:            return "Publish"
        case .markShipped:        return "Mark shipped"
        case .endListing:         return "End listing"
        case .dropPrice(let pct): return "Drop -\(pct)%"
        case .grade:              return "Grade"
        case .exportCSV:          return "Export CSV"
        }
    }

    public var systemImage: String {
        switch self {
        case .createDraft: return "doc.text"
        case .publish:     return "paperplane.fill"
        case .markShipped: return "shippingbox.fill"
        case .endListing:  return "stop.circle"
        case .dropPrice:   return "arrow.down.circle"
        case .grade:       return "checkmark.seal"
        case .exportCSV:   return "square.and.arrow.up"
        }
    }

    public var isDestructive: Bool {
        switch self {
        case .endListing: return true
        default:          return false
        }
    }

    /// Confirmation copy phrased per-action. `count` slots in the
    /// number-of-rows the action will touch.
    public func confirmationTitle(count: Int) -> String {
        let suffix = count == 1 ? "item" : "items"
        switch self {
        case .createDraft:        return "Create \(count) \(suffix) as drafts?"
        case .publish:            return "Publish \(count) \(suffix) to eBay?"
        case .markShipped:        return "Mark \(count) \(suffix) as shipped?"
        case .endListing:         return "End \(count) \(suffix)?"
        case .dropPrice(let pct): return "Drop price -\(pct)% on \(count) \(suffix)?"
        case .grade:              return "Grade \(count) \(suffix)?"
        case .exportCSV:          return "Export \(count) \(suffix) as CSV?"
        }
    }

    /// Stage-appropriate action sets, mirroring src/pages/flipdesk/listings.tsx
    /// bottom-bar predicate.
    public static func actions(for stage: InventoryStage) -> [BulkAction] {
        switch stage {
        case .unlisted:
            // Unlisted holds undrafted and drafted rows together, so it offers
            // both halves: Create draft for the first, Publish for the second.
            // .grade IS surfaced: InventoryListView intercepts it into the
            // dedicated bulk-grade sheet (it never hits the executor path).
            return [.publish, .grade, .createDraft, .exportCSV]
        case .active:
            return [.dropPrice(percent: 10), .endListing, .exportCSV]
        case .sold:
            return [.markShipped, .exportCSV]
        case .shipped:
            return [.exportCSV]
        case .returned:
            return [.exportCSV]
        case .all:
            // Mixed-status selection — restrict to non-destructive
            // actions that work uniformly. Per-action validation in
            // BulkActionExecutor still kicks in.
            return [.exportCSV]
        }
    }
}

/// Outcome of executing a bulk action across a set of items. Caller
/// renders `summary` as a toast and `failures` in a follow-up list when
/// non-empty.
public struct BulkActionResult: Equatable {
    public let action: BulkAction
    public let succeeded: Int
    public let failures: [Failure]
    /// US-1506: non-fatal notices — e.g. a listing that ended in FlipDesk but
    /// whose eBay withdraw didn't actually happen (already inactive / no linked
    /// offer). These items still count toward `succeeded`; the warning tells the
    /// seller the eBay side may be unchanged so they can verify.
    public var warnings: [String] = []

    public struct Failure: Equatable {
        public let itemId: String
        public let message: String
    }

    public var total: Int { succeeded + failures.count }

    public var summary: String {
        let suffix = total == 1 ? "item" : "items"
        let base: String
        if failures.isEmpty {
            base = "Updated \(succeeded) \(suffix)."
        } else if succeeded == 0 {
            base = "All \(failures.count) \(suffix) failed."
        } else {
            base = "Updated \(succeeded) of \(total) \(suffix); \(failures.count) failed."
        }
        if warnings.isEmpty { return base }
        // Surface a single representative notice inline; the caller can render
        // the full `warnings` list if it wants more detail.
        return "\(base) \(warnings.count == 1 ? warnings[0] : "\(warnings.count) listings ended in FlipDesk only — verify on eBay.")"
    }
}

File: android/app/src/main/java/com/gradethread/app/inventory/BulkAction.kt
package com.gradethread.app.inventory

import com.gradethread.app.R
import com.gradethread.app.ui.UiMessage

/**
 * US-1348: one action offered in the multi-select bar.
 *
 * DELIBERATELY NARROWER THAN iOS. The iOS set includes Publish, End listing
 * and the eBay-backed price drop.
 *
 * The original reason given here — "Android has no listing composer yet" —
 * stopped being true when PublishSheet was wired into the item canvas, and
 * US-2490 has since added reprice, resubmit and end as PER-LISTING actions on
 * the listing card. What remains true is the reason to keep them off the BULK
 * bar: each of these pushes to a live marketplace, and a multi-select that
 * ends forty listings on one tap is a different risk from ending one after a
 * confirmation. Bulk versions want the same treatment the web gives them
 * (selection, confirm, per-row result), which is its own story.
 */
sealed class BulkAction {

    /** Move the selection to `drafted`. */
    object CreateDraft : BulkAction()

    object MarkShipped : BulkAction()

    /** Reduce TARGET price by [percent] — the local price, not a live listing. */
    data class DropPrice(val percent: Int) : BulkAction()

    /**
     * Grading is intercepted by the list, not run here: it needs a tier,
     * readiness and credits, which is the US-1339 sheet's whole job.
     */
    object Grade : BulkAction()

    object Delete : BulkAction()

    val id: String
        get() = when (this) {
            CreateDraft -> "create_draft"
            MarkShipped -> "mark_shipped"
            is DropPrice -> "drop_price_$percent"
            Grade -> "grade"
            Delete -> "delete"
        }

    val label: UiMessage
        get() = when (this) {
            CreateDraft -> UiMessage(R.string.bulk_label_create_draft)
            MarkShipped -> UiMessage(R.string.bulk_label_mark_shipped)
            is DropPrice -> UiMessage(R.string.bulk_label_drop_price, args = listOf(percent))
            Grade -> UiMessage(R.string.bulk_label_grade)
            Delete -> UiMessage(R.string.bulk_label_delete)
        }

    /** A plurals resource whose count is also the number in the sentence. */
    private fun plural(res: Int, count: Int) = UiMessage(res, args = listOf(count), quantity = count)

    /** Destructive actions confirm before running. */
    val destructive: Boolean get() = this == Delete

    /**
     * Whether the action can be taken back within the undo window.
     *
     * Delete is NOT reversible: the rows are gone server-side and their photos
     * cascade with them, so an "Undo" that couldn't restore the images would
     * be a promise we can't keep. It confirms up front instead.
     */
    val reversible: Boolean
        get() = when (this) {
            CreateDraft, MarkShipped -> true
            is DropPrice -> true
            Grade, Delete -> false
        }

    /**
     * US-2976: a plurals resource per action, with the COUNT as both the
     * selector and the number in the sentence.
     *
     * The old `if (count == 1) "item" else "items"` is exactly the shape that
     * does not survive translation - Spanish agrees the verb and the adjective
     * with the noun too, so "Marcar 1 artículo como enviado" and "Marcar 3
     * artículos como enviados" differ in three places, not one.
     */
    fun confirmationTitle(count: Int): UiMessage = when (this) {
        CreateDraft -> plural(R.plurals.bulk_confirm_create_draft, count)
        MarkShipped -> plural(R.plurals.bulk_confirm_mark_shipped, count)
        is DropPrice -> UiMessage(
            R.plurals.bulk_confirm_drop_price,
            args = listOf(count, percent),
            quantity = count,
        )

        Grade -> plural(R.plurals.bulk_confirm_grade, count)
        Delete -> plural(R.plurals.bulk_confirm_delete, count)
    }

    companion object {

        /**
         * Stage-appropriate sets, mirroring the iOS bottom-bar predicate minus
         * the eBay-backed actions.
         *
         * A MIXED selection (the All tab) gets nothing but Delete and Grade,
         * because a status action across mixed statuses either regresses some
         * rows or silently skips them — and both read as a bug.
         */
        fun forStage(stage: InventoryStage): List<BulkAction> = when (stage) {
            InventoryStage.TO_LIST -> listOf(Grade, CreateDraft, Delete)
            InventoryStage.DRAFTS -> listOf(Grade, Delete)
            InventoryStage.ACTIVE -> listOf(DropPrice(10), Delete)
            InventoryStage.SOLD -> listOf(MarkShipped, Delete)
            InventoryStage.SHIPPED -> listOf(Delete)
            InventoryStage.RETURNED -> listOf(Delete)
            InventoryStage.ALL -> listOf(Grade, Delete)
        }
    }
}

/** What a batch did. */
data class BulkActionResult(val action: BulkAction, val succeeded: Int, val failures: List<Failure> = emptyList()) {
    /**
     * US-2976: [message] is a UiMessage because the two sources of this
     * sentence are different. A skip we decided ("No target price to drop.")
     * is our copy and translates; a server rejection is the server's own words
     * and rides as `detail`. Flattening both to String made the second look
     * like the first.
     */
    data class Failure(val itemId: String, val message: UiMessage)

    val total: Int get() = succeeded + failures.size

    /**
     * The summary line.
     *
     * Partial success is named explicitly — "Updated 7 of 9" — because a batch
     * that half-worked and reported "Done" is how a seller discovers two
     * unshipped orders a week later.
     */
    val summary: UiMessage
        get() = when {
            failures.isEmpty() -> UiMessage(
                R.plurals.bulk_result_updated,
                args = listOf(succeeded),
                quantity = succeeded,
            )

            succeeded == 0 -> UiMessage(
                R.plurals.bulk_result_all_failed,
                args = listOf(failures.size),
                quantity = failures.size,
            )

            // The partial case pluralises on the TOTAL, which is the noun the
            // sentence is about - "Updated 1 of 9 items" is nine items, not one.
            else -> UiMessage(
                R.plurals.bulk_result_partial,
                args = listOf(succeeded, total, failures.size),
                quantity = total,
            )
        }

    val hasFailures: Boolean get() = failures.isNotEmpty()
}

/**
 * What an undo needs to put back.
 *
 * Snapshots are taken BEFORE the batch runs, and only for the items that
 * actually succeeded — reverting an item whose update failed would write a
 * value it never had.
 */
data class BulkUndo(
    val label: UiMessage,
    val statuses: Map<String, String> = emptyMap(),
    val targetPrices: Map<String, Double?> = emptyMap(),
) {
    val isEmpty: Boolean get() = statuses.isEmpty() && targetPrices.isEmpty()

    companion object {
        /** How long the undo affordance stays up (iOS US-972). */
        const val WINDOW_SECONDS = 6
    }
}

object BulkPricing {

    /**
     * Apply a percentage drop.
     *
     * Rounded to whole cents and floored at one cent: a percentage of a small
     * price can otherwise land on 0, and an item listed at nothing is worse
     * than one that didn't move. Returns null when there is no price to drop,
     * so the item fails with a reason rather than silently gaining one.
     */
    fun dropped(current: Double?, percent: Int): Double? {
        if (current == null || current <= 0.0) return null
        val next = current * (100 - percent) / 100.0
        return (Math.round(next * 100) / 100.0).coerceAtLeast(0.01)
    }
}
